Core/ALife.Core/WorldObjects/Agents/Brains/NeuralNetworkBrain.cs

```diff
@@ -28,217 +31,318 @@ public NeuralNetworkBrain(Agent self, List<int> layers)
+        public NeuralNetworkBrain(Agent self, double modificationRate, double mutabilityRate, List<int> hiddenLayerNeuronCounts)
         {
-            if(layerNeuronCounts.Count < 1) throw new ArgumentOutOfRangeException("Not enough layers for a Neural Network Brain");
-
+            if(hiddenLayerNeuronCounts is null
+               || hiddenLayerNeuronCounts.Count < 1)
+            {
+                throw new ArgumentOutOfRangeException("Not enough layers for a Neural Network Brain");
+            }
             this.self = self;
-            ModificationRate = modificationRate;
-            MutabilityRate = mutabilityRate;
+            this.ModificationRate = 0.70;
+            this.MutabilityRate = 0.005;
 
-            //First we initialize the Top Layer to have all the inputs available.
-            Layer senseLayer = CreateSenseLayer(self);
-            Layers.Add(senseLayer);
-
-            //Next we implement all the middle layers
-            for(int i = 0; i < layerNeuronCounts.Count; i++)
-            {
-                Layer newLayer = new Layer(layerNeuronCounts[i]);
-                Layers.Add(newLayer);
+            this.Layers = NeuralNetworkBrainFactory.CreateRandomNeuralNetwork(self, hiddenLayerNeuronCounts);
+            this.actions = Layers[Layers.Count - 1];
+        }
 
-                //This pulls the layer above us. Layer 0 is the sense layer. Layer 1 is the first created hidden layer.
-                List<Neuron> aboveLayerNeurons = Layers[i].Neurons;
+        public NeuralNetworkBrain(Agent self, NeuralNetworkBrain templateBrain, bool exactCopy)
+        {
+            this.self = self;
+            this.ModificationRate = templateBrain.ModificationRate;
+            this.MutabilityRate = templateBrain.MutabilityRate;
 
-                for(int n = 0; n < layerNeuronCounts[i]; n++)
-                {
-                    //Add a neuron
-                    Neuron neu = new Neuron("HN:" + (i + 1) + "." + (n + 1));
-                    newLayer.Neurons.Add(neu);
-                    for(int d = 0; d < aboveLayerNeurons.Count; d++)
-                    {
-                        neu.UpstreamDendrites.Add(new Dendrite(aboveLayerNeurons[d]));
-                    }
-                }
+            if(exactCopy)
+            {
+                this.Layers = NeuralNetworkBrainFactory.CreateClonedNeuralNetwork(self, templateBrain);
             }
-
-            Layer actionLayer = CreateActionLayer(self, Layers[Layers.Count - 1]);
-            Layers.Add(actionLayer);
-            actions = actionLayer;
+            else
+            {
+                this.Layers = NeuralNetworkBrainFactory.CreateEvolvedNeuralNetwork(self, templateBrain);
+            }
+            
+            this.actions = Layers[Layers.Count - 1];
         }
 
-        private NeuralNetworkBrain(Agent self, NeuralNetworkBrain templateBrain, bool exactCopy)
+        public NeuralNetworkBrain(Agent self, string inputString)
```

Comment:
Add a todo?

Or make private and have brains in this scenario initialized with a distinct constructor method.

Reply:
This one needs to be public currently.